samza-core/src/main/java/org/apache/samza/table/remote/RemoteReadWriteTable.java

```diff
@@ -69,33 +68,36 @@ public void init(SamzaContainerContext containerContext, TaskContext taskContext
     super.init(containerContext, taskContext);
     writeMetrics = new DefaultTableWriteMetrics(containerContext, taskContext, this, tableId);
     TableMetricsUtil tableMetricsUtil = new TableMetricsUtil(containerContext, taskContext, this, tableId);
-    putThrottleNs = tableMetricsUtil.newTimer("put-throttle-ns");
+    writeThrottler.setTimerMetric(tableMetricsUtil.newTimer("put-throttle-ns"));
   }
 
   /**
    * {@inheritDoc}
    */
   @Override
   public void put(K key, V value) {
+    try {
+      putAsync(key, value).get();
+    } catch (Exception e) {
+      throw new SamzaException(e);
+    }
+  }
 
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public CompletableFuture<Void> putAsync(K key, V value) {
+    Preconditions.checkNotNull(key);
     if (value == null) {
-      delete(key);
-      return;
+      return deleteAsync(key);
     }
 
-    try {
-      writeMetrics.numPuts.inc();
-      if (rateLimitWrites) {
-        throttle(key, value, RL_WRITE_TAG, writeCreditFn, putThrottleNs);
-      }
-      long startNs = System.nanoTime();
-      writeFn.put(key, value);
-      writeMetrics.putNs.update(System.nanoTime() - startNs);
-    } catch (Exception e) {
-      String errMsg = String.format("Failed to put a record, key=%s, value=%s", key, value);
-      logger.error(errMsg, e);
-      throw new SamzaException(errMsg, e);
-    }
+    writeMetrics.numPuts.inc();
+    return asyncHelper.execute(key, value, writeFn::putAsync, writeMetrics.putNs)
+        .exceptionally(e -> {
+            throw new SamzaException(String.format("Failed to put a record, key=%s, value=%s", key, value), e);
```

Comment:
I tend to think we should give the developer all information to debug the errors in case. Logging value is useful when it is the reason for the failed request. Given such errors are exceptional case, so I don't expect us logging it excessively. Typically the value is a avro record with human-readable information. 

I don't have a strong opinion so I'm open to remove too it too. Let me know.

Reply:
Another concern is that the value may contain PII, typically with key one can trace back to the value from input streams.

Reply:
Sure, I'll only log the key part then.